stl/inc/limits

```diff
@@ -1009,24 +1011,96 @@ public:
+    return static_cast<int>(_Nn) - static_cast<int>(_Val);
+}
 
-template <class _Ty, enable_if_t<_Is_standard_unsigned_integer<_Ty>, int> = 0>
-_NODISCARD constexpr int _Countr_zero(const _Ty _Val) noexcept {
-    if (_Val == 0) {
-        return numeric_limits<_Ty>::digits;
-    }
+// Implementation of countr_zero without using specialized CPU instructions.
+// Used at compile time and when said instructions are not supported.
+// see "Hacker's Delight" section 5-4
+template <class _Ty>
+_NODISCARD constexpr int _Countr_zero_helper(const _Ty _Val) noexcept {
+    constexpr int _Digits = std::numeric_limits<_Ty>::digits;
+    return _Digits - _Countl_zero_helper(static_cast<_Ty>(static_cast<_Ty>(~_Val) & static_cast<_Ty>(_Val - 1)));
+}
+
+#if defined(_M_IX86) || defined(_M_X64)
+
+extern "C" {
+extern int __isa_available;
+#ifdef __clang__
+#define _TZCNT_U32 __builtin_ia32_tzcnt_u32
+#define _TZCNT_U64 __builtin_ia32_tzcnt_u64
+#else // ^^^ __clang__ / ! __clang__ vvv
+__MACHINEX86_X64(unsigned int _tzcnt_u32(unsigned int))
+__MACHINEX64(unsigned __int64 _tzcnt_u64(unsigned __int64));
+#define _TZCNT_U32 _tzcnt_u32
+#define _TZCNT_U64 _tzcnt_u64
+#endif // __clang__
+}
 
-    if constexpr (sizeof(_Ty) <= sizeof(unsigned int)) {
-        return __builtin_ctz(_Val);
+template <class _Ty>
+int _Checked_x86_x64_countr_zero(const _Ty _Val) noexcept {
+    const bool _Have_tzcnt = __isa_available >= __ISA_AVAILABLE_AVX2;
+    constexpr int _Digits  = numeric_limits<_Ty>::digits;
+    constexpr _Ty _Max     = (numeric_limits<_Ty>::max)();
+    if (!_Have_tzcnt && _Val == 0) {
+        return _Digits;
+    }
+
+    if constexpr (_Digits <= 32) {
+        // Intended widening to int. This operation means that a narrow 0 will widen
+        // to 0xFFFF....FFFF0... instead of 0. We need this to avoid counting all the zeros
+        // of the wider type
+        return static_cast<int>(_TZCNT_U32(static_cast<unsigned int>(~_Max | _Val)));
     } else {
-        return __builtin_ctzll(_Val);
+#ifdef _M_IX86
+        unsigned int _High = _Val >> 32;
+        unsigned int _Low  = static_cast<unsigned int>(_Val);
+        int _Result        = _Checked_x86_x64_countr_zero(_Low);
+        _Result += !_Low ? _Checked_x86_x64_countr_zero(_High) : 0;
+        return _Result;
+#else // ^^^ _M_IX86 / !_M_IX86 vvv
+        return static_cast<int>(_TZCNT_U64(_Val));
+#endif // _M_IX86
     }
 }
-#endif // __cpp_lib_bitops
+#undef _TZCNT_U32
+#undef _TZCNT_U64
+#endif // defined(_M_IX86) || defined(_M_X64)
+
+template <class _Ty>
+constexpr bool _Is_standard_unsigned_integer =
+    _Is_any_of_v<remove_cv_t<_Ty>, unsigned char, unsigned short, unsigned int, unsigned long, unsigned long long>;
+
+template <class _Ty, enable_if_t<_Is_standard_unsigned_integer<_Ty>, int> = 0>
+_NODISCARD constexpr int _Countr_zero(const _Ty _Val) noexcept {
+#ifdef __cpp_lib_is_constant_evaluated
+    if (!_STD is_constant_evaluated())
+#endif // __cpp_lib_is_constant_evaluated

```

Comment:
I believe that this code pattern will emit "unreachable code" warnings (we globally suppress C4702 but I still recall this in particular being a problem - maybe it was for Clang when the system header suppression is disabled). When `__cpp_lib_is_constant_evaluated` isn't defined, and we're x86/x64, this function contains two `return` statements, the second of which is clearly unreachable.

Can we restructure this logic?

Reply:
This is not yet resolved.